src/interpreter.c

```diff
@@ -55,18 +64,132 @@ JL_DLLEXPORT jl_value_t *jl_interpret_toplevel_expr_in(jl_module_t *m,
+                }
+                assert(jl_is_symbol(var));
+                int k;
+                for (k = 0; k < s.nl; k++) {
+                    if (s.locals[2*k] == var) {
+                        ref_index[i] = 1 + 2*k + 1;
+                        val = s.locals[2*k + 1];
+                        break;
+                    }
+                }
+                assert(k < s.nl);
+            }
+            jl_value_t *param = NULL, *ty = NULL, *ref = NULL;
+            JL_GC_PUSH3(&param, &ty, &ref);
+            param = jl_alloc_svec(1);
+            jl_svecset(param, 0, jl_typeof(val));
+            int sz = jl_datatype_size(jl_typeof(val));
+            refs[i] = malloc(sz);
+            memcpy(refs[i], val, sz);
+            argv[i] = jl_new_struct_uninit(jl_apply_type(jl_pointer_type, param));
+            *(void**)argv[i] = refs[i];
+            JL_GC_POP();
+        } else { /* sane path */
+            argv[i] = eval(arg, s);
+            ref_index[i] = 0;
+        }
+    }
+    /* nothing to see here */
     jl_value_t *result = jl_apply_generic(argv, nargs);
+    for (int k = 0; k < nargs; k++) {
+        if (ref_index[k] != 0) {
+            int ri = ref_index[k] - 1;
+            jl_value_t *ty = jl_typeof(s.locals[ri]);
+            s.locals[ri] = jl_new_struct_uninit(ty);
+            memcpy(s.locals[ri], refs[k], jl_datatype_size(ty));
+            free(refs[k]);
+        }
+    }
     JL_GC_POP();
     return result;
 }
 
+jl_value_t *jl_apply_interpreted(jl_value_t **argv, uint32_t nargs)
+{
+    if (jl_typeof(argv[0]) == jl_intrinsic_type) {
+        int intr_id = *(int*)argv[0];
+        if (intr_id == 89) { // ccall implementation
+            jl_value_t *gen_ccall = jl_eval_string("Base.generic_ccall");
+            argv[0] = gen_ccall;
+            if (!jl_is_cpointer(argv[1])) {
+                argv[1] = jl_native_sym(argv[1], jl_voidpointer_type);
+            }
+            argv[3] = jl_inst_concrete_tupletype(argv[3]);
+            assert(nargs >= 4);
+            int was_enabled = jl_interpreter_enabled;
+            jl_interpreter_enabled = 0; // avoid recursing in there while generating generic_ccall's body
+            jl_value_t *result;
+            if (nargs == 4) {
+                jl_value_t **argv2 = alloca(sizeof(jl_value_t*)*(nargs+1));
+                memcpy(argv2, argv, nargs*sizeof(jl_value_t*));
+                argv2[4] = jl_emptysvec;
+                result = jl_apply_generic_ex(argv2, 5, 0);
+            } else {
+                jl_value_t *ccall_args = jl_alloc_svec(nargs - 4); // not rooted until next assignment
+                memcpy(jl_svec_data(ccall_args), &argv[4], (nargs - 4)*sizeof(jl_value_t*));
+                argv[4] = ccall_args;
+                result = jl_apply_generic_ex(argv, 5, 0);
+            }
+            jl_interpreter_enabled = was_enabled;
+            return result;
+        } else if (intr_id == 90) { // cglobal implementation
```

Comment:
ah, foey: i never pushed this fix to master: c0dc384 (minus the diff for src/toplevel.c)